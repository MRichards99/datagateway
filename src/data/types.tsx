export interface InvestigationData {
  ID: string;
  TITLE: string;
  VISIT_ID: number;
  RB_NUMBER: string;
  DOI: string;
  SIZE: number;
  INSTRUMENT: { NAME: string };
  STARTDATE: Date;
  ENDDATE: Date;
  [key: string]: string | number | Date | { NAME: string };
}

export interface DatasetData {
  ID: string;
  NAME: string;
  SIZE: number;
  CREATE_TIME: Date;
  MOD_TIME: Date;
  [key: string]: string | number | Date;
}

export interface DatafileData {
  ID: string;
  NAME: string;
  LOCATION: string;
  SIZE: number;
  MOD_TIME: Date;
  [key: string]: string | number | Date;
}

export interface FoodData {
  id: number;
  dessert: string;
  calories: number;
  fat: number;
  carbs: number;
  protein: number;
  [key: string]: string | number;
}

export type EntityType =
  | InvestigationData
  | DatasetData
  | DatafileData
  | FoodData;
