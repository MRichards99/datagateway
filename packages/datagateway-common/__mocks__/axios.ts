import axios from 'axios';

export default {
  get: jest.fn(() => Promise.resolve({ data: {} })),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  delete: jest.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
};