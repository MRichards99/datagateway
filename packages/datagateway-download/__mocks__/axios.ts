import axios from 'axios';

export default {
  get: jest.fn(path => {
    if (path.includes('/topcat/user/cart/')) {
      return Promise.resolve({ data: { cartItems: [] } });
    } else if (path === '/datagateway-download-settings.json') {
      return Promise.resolve({
        data: {
          facilityName: 'LILS',
          apiUrl: 'http://scigateway-preprod.esc.rl.ac.uk:5000',
          downloadApiUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/topcat',
          idsUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/ids',
          accessMethods: {
            https: {
              idsUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/ids',
              displayName: 'HTTPS',
              description: 'Example description for HTTPS access method.',
            },
            globus: {
              idsUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/ids',
              displayName: 'Globus',
              description: 'Example description for Globus access method.',
            },
          },
        },
      });
    } else {
      return Promise.resolve({ data: {} });
    }
  }),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
};
