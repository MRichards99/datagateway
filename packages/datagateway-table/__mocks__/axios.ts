import axios from 'axios';

export default {
  get: jest.fn(path => {
    if (path === '/settings.json') {
      return Promise.resolve({
        data: {
          'ui-strings': '/res/default.json',
        },
      });
    } else {
      return Promise.resolve({
        data: {},
      });
    }
  }),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
};
