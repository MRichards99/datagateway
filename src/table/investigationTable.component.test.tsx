import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import InvestigationTable from './investigationTable.component';
import { initialState } from '../state/reducers/dgtable.reducer';
import configureStore from 'redux-mock-store';
import { StateType } from '../state/app.types';
import {
  fetchInvestigationsRequest,
  filterTable,
  sortTable,
} from '../state/actions';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { TableSortLabel } from '@material-ui/core';
import Table from './table.component';
import { MemoryRouter } from 'react-router';
import axios from 'axios';

describe('Investigation table component', () => {
  let shallow;
  let mount;
  let mockStore;
  let state: StateType;
  (axios.get as jest.Mock).mockImplementation(() =>
    Promise.resolve({ data: [] })
  );

  beforeEach(() => {
    shallow = createShallow({ untilSelector: 'div' });
    mount = createMount();

    mockStore = configureStore([thunk]);
    state = JSON.parse(JSON.stringify({ dgtable: initialState }));
    state.dgtable.data = [
      {
        ID: 1,
        TITLE: 'Test 1',
        VISIT_ID: '1',
        RB_NUMBER: '1',
        DOI: 'doi 1',
        SIZE: 1,
        INSTRUMENT: {
          NAME: 'LARMOR',
        },
        STARTDATE: '2019-07-23',
        ENDDATE: '2019-07-24',
      },
    ];
  });

  afterEach(() => {
    mount.cleanUp();
  });

  it('renders correctly', () => {
    const wrapper = shallow(<InvestigationTable store={mockStore(state)} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('sends fetchInvestigations action on load', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <InvestigationTable />
        </MemoryRouter>
      </Provider>
    );

    expect(testStore.getActions()[0]).toEqual(fetchInvestigationsRequest());
  });

  it('sends filterTable action on text filter', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <InvestigationTable />
        </MemoryRouter>
      </Provider>
    );

    const filterInput = wrapper.find('input').first();
    filterInput.instance().value = 'test';
    filterInput.simulate('change');

    expect(testStore.getActions()[1]).toEqual(filterTable('TITLE', 'test'));

    filterInput.instance().value = '';
    filterInput.simulate('change');

    expect(testStore.getActions()[2]).toEqual(filterTable('TITLE', null));
  });

  it('sends filterTable action on date filter', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <InvestigationTable />
        </MemoryRouter>
      </Provider>
    );

    const filterInput = wrapper.find('input').last();
    filterInput.instance().value = '2019-08-06';
    filterInput.simulate('change');

    expect(testStore.getActions()[1]).toEqual(
      filterTable('ENDDATE', { endDate: '2019-08-06' })
    );

    filterInput.instance().value = '';
    filterInput.simulate('change');

    expect(testStore.getActions()[2]).toEqual(filterTable('ENDDATE', null));
  });

  it('sends sortTable action on sort', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <InvestigationTable />
        </MemoryRouter>
      </Provider>
    );

    wrapper
      .find(TableSortLabel)
      .first()
      .simulate('click');

    expect(testStore.getActions()[1]).toEqual(sortTable('TITLE', 'asc'));
  });

  it('renders details panel correctly', () => {
    const wrapper = shallow(
      <MemoryRouter>
        <InvestigationTable store={mockStore(state)} />
      </MemoryRouter>
    );
    const detailsPanelWrapper = shallow(
      wrapper.find(Table).prop('detailsPanel')(state.dgtable.data[0])
    );
    expect(detailsPanelWrapper).toMatchSnapshot();
  });

  it('renders investigation title as a link', () => {
    const wrapper = mount(
      <Provider store={mockStore(state)}>
        <MemoryRouter>
          <InvestigationTable />
        </MemoryRouter>
      </Provider>
    );

    expect(
      wrapper
        .find('[aria-colindex=2]')
        .find('p')
        .children()
    ).toMatchSnapshot();
  });

  it('renders file size as bytes', () => {
    const wrapper = mount(
      <Provider store={mockStore(state)}>
        <MemoryRouter>
          <InvestigationTable />
        </MemoryRouter>
      </Provider>
    );

    expect(
      wrapper
        .find('[aria-colindex=6]')
        .find('p')
        .text()
    ).toEqual('1 B');
  });
});
