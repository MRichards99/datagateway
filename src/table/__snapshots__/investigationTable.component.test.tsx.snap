// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Investigation table component renders correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "height": 768,
      "width": "100%",
    }
  }
>
  <WithStyles(VirtualizedTable)
    columns={
      Array [
        Object {
          "cellContentRenderer": [Function],
          "dataKey": "TITLE",
          "filterComponent": [Function],
          "label": "Title",
        },
        Object {
          "dataKey": "VISIT_ID",
          "filterComponent": [Function],
          "label": "Visit ID",
        },
        Object {
          "dataKey": "RB_NUMBER",
          "filterComponent": [Function],
          "label": "RB Number",
        },
        Object {
          "dataKey": "DOI",
          "filterComponent": [Function],
          "label": "DOI",
        },
        Object {
          "cellContentRenderer": [Function],
          "dataKey": "SIZE",
          "label": "Size",
        },
        Object {
          "dataKey": "INSTRUMENT.NAME",
          "filterComponent": [Function],
          "label": "Instrument",
        },
        Object {
          "dataKey": "STARTDATE",
          "label": "Start Date",
        },
        Object {
          "dataKey": "ENDDATE",
          "label": "End Date",
        },
      ]
    }
    data={
      Array [
        Object {
          "DOI": "doi 1",
          "ENDDATE": "2019-07-24",
          "ID": 1,
          "INSTRUMENT": Object {
            "NAME": "LARMOR",
          },
          "RB_NUMBER": "1",
          "SIZE": 1,
          "STARTDATE": "2019-07-23",
          "TITLE": "Test 1",
          "VISIT_ID": "1",
        },
      ]
    }
    detailsPanel={[Function]}
    onSort={[Function]}
    sort={Object {}}
  />
</div>
`;

exports[`Investigation table component renders details panel correctly 1`] = `
<div>
  <WithStyles(ForwardRef(Typography))>
    <b>
      Proposal: 
    </b>
    1
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))>
    <b>
      Title: 
    </b>
    Test 1
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))>
    <b>
      Start Date: 
    </b>
    2019-07-23
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))>
    <b>
      End Date: 
    </b>
    2019-07-24
  </WithStyles(ForwardRef(Typography))>
</div>
`;

exports[`Investigation table component renders investigation title as a link 1`] = `
<Link
  to="/browse/investigation/1/dataset"
>
  <a
    href="/browse/investigation/1/dataset"
    onClick={[Function]}
  >
    Test 1
  </a>
</Link>
`;
